src/vmemcache.c

```diff
@@ -78,6 +237,149 @@ int
 vmemcache_put(VMEMcache *cache, const char *key, size_t ksize,
 				const char *value, size_t value_size)
 {
+	struct cache_entry *entry;
+	struct heap_entry he;
+
+	entry = Zalloc(sizeof(struct cache_entry) + ksize);
+	if (entry == NULL) {
+		ERR("!Zalloc");
+		return -1;
+	}
+
+	util_mutex_init(&entry->value.lock);
+	util_mutex_lock(&entry->value.lock);
+
+	entry->key.ksize = ksize;
+	memcpy(entry->key.key, key, ksize);
+
+	entry->value.p_entry =
+		cache->repl.ops->repl_p_insert(cache->repl.head, entry);
+
+	size_t left_to_allocate = value_size;
+
+	while (left_to_allocate > 0) {
+		he = vmcache_alloc(cache->heap, left_to_allocate);
+		if (HEAP_ENTRY_IS_NULL(he)) {
+			if (vmemcache_evict(cache, NULL, 0)) {
+				errno = ENOMEM;
+				goto error_no_mem;
+			}
+			continue;
+		}
+
+		if (VEC_PUSH_BACK(&entry->value.fragments, he))
+			FATAL("out of system memory");
+
+		if (left_to_allocate <= he.size)
+			break;
+
+		left_to_allocate -= he.size;
+	}
+
+	vmemcache_populate_fragments(entry, value, value_size);
+	entry->value.vsize = value_size;
+
+	ravl_insert(cache->index, entry);
+
+	util_mutex_unlock(&entry->value.lock);
+
+	return 0;
+
+error_no_mem:
+	VEC_FOREACH(he, &entry->value.fragments) {
+		vmcache_free(cache->heap, he);
+	}
+
+	VEC_DELETE(&entry->value.fragments);
+
+	util_mutex_unlock(&entry->value.lock);
+
+	return -1;
+}
+
+/*
+ * vmemcache_memcpy_fragments -- (internal) memcpy fragments
+ */
+static size_t
+vmemcache_memcpy_fragments(char *vbuf, size_t vbufsize, size_t offset,
+				struct cache_entry *entry)
+{
+	struct heap_entry he;
+	size_t copied = 0;
+
+	VEC_FOREACH(he, &entry->value.fragments) {
+		if (offset > he.size) {
+			offset -= he.size;
+			continue;
+		}
+
+		size_t off = 0;
+		size_t len = he.size;
+
+		if (offset > 0) {
+			off += offset;
+			len -= offset;
+			offset = 0;
+		}
+
+		if (len > vbufsize)
+			len = vbufsize;
+
+		memcpy(vbuf, (char *)he.ptr + off, len);
+
+		vbufsize -= len;
+		vbuf += len;
+		copied += len;
+
+		if (vbufsize == 0)
+			return copied;
+	}
+
+	return copied;
+}
+
+/*
+ * vmemcache_find_key - find the key in the index
+ *                     and optionally remove it from the index
+ */
+static int
+vmemcache_find_key_remove(VMEMcache *cache, const char *key, size_t ksize,
+			struct cache_entry **entry, int remove)
+{
+	*entry = NULL;
+
+	struct cache_entry *e = Zalloc(sizeof(struct cache_entry) + ksize);
+	if (e == NULL) {
+		ERR("!Zalloc");
+		return -1;
+	}
+
+	e->key.ksize = ksize;
+	memcpy(e->key.key, key, ksize);
+
+	struct ravl_node *node;
+	node = ravl_find(cache->index, e, RAVL_PREDICATE_EQUAL);
+	Free(e);
+	if (node == NULL) {
+		LOG(1, "cannot find an element with the given key");
+		return 0;
+	}
+
+	e = ravl_data(node);
+
+	if (remove &&
+	    __sync_bool_compare_and_swap(&e->value.removed_from_index, 0, 1)) {
```

Comment:
I know I told you to add the "removed" value, but I think just the refcount is enough ;)

Reply:
Done